libraries/chain/market_evaluator.cpp

```diff
@@ -134,6 +134,93 @@ object_id_type limit_order_create_evaluator::do_apply(const limit_order_create_o
    return order_id;
 } FC_CAPTURE_AND_RETHROW( (op) ) }
 
+void_result limit_order_update_evaluator::do_evaluate(const limit_order_update_operation& o)
+{ try {
+   const database& d = db();
+   FC_ASSERT(d.head_block_time() >= HARDFORK_CORE_1604_TIME, "Operation has not activated yet");
+   _order = &o.order(d);
+
+   // Check this is my order
+   FC_ASSERT(o.seller == _order->seller, "Cannot update someone else's order");
+
+   // Check new price is compatible, and determine whether it becomes the best offer on the market
+   if (o.new_price) {
+      auto base_id = o.new_price->base.asset_id;
+      auto quote_id = o.new_price->quote.asset_id;
+      FC_ASSERT(base_id == _order->sell_price.base.asset_id && quote_id == _order->sell_price.quote.asset_id,
+                "Cannot update limit order with incompatible price");
+      const auto& order_index = d.get_index_type<limit_order_index>().indices().get<by_price>();
+      auto top_of_book = order_index.upper_bound(boost::make_tuple(price::max(base_id, quote_id)));
+      FC_ASSERT(top_of_book->sell_price.base.asset_id == base_id && top_of_book->sell_price.quote.asset_id == quote_id,
+                "Paradox: attempting to update an order in a market that has no orders? There's a logic error somewhere.");
+
+      // If the new price of our order is greater than the price of the order at the top of the book, we should match orders at the end.
+      // Otherwise, we can skip matching because there's no way this change could trigger orders to fill.
+      should_match_orders = (*o.new_price > top_of_book->sell_price);
+   }
```

Comment:
There is no check for valid range of price change, so the new price can be the same as the original price.

Actually, this allows the order owner to slightly update the price so that this order is ahead or after another (for competition or safety), while receiving the same amount if it ends up being filled. This was not possible with cancel + recreate. I'm not sure if it's good or not. Maybe it leads to more chain activity. Is chain bloat a real problem?